File: sdks/ts/packages/golem-ts-sdk/tests/retry.test.ts
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Golem Source License v1.1 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import {
  Duration,
  NamedPolicy,
  Policy,
  Predicate,
  Props,
  toRawDuration,
  toRawNamedPolicy,
  toRawPredicateValue,
} from '../src/host/retryBuilder';

describe('retry builder', () => {
  it('compiles predicates and policies into flattened raw WIT shapes', () => {
    const appliesWhen = Predicate.eq(Props.verb, 'get').and(Predicate.gte(Props.statusCode, 500));
    const policy = NamedPolicy.named(
      'http-errors',
      Policy.exponential(Duration.seconds(1), 2)
        .maxRetries(3)
        .withJitter(0.25)
        .onlyWhen(Predicate.matchesGlob(Props.errorType, 'timeout*')),
    )
      .priority(10)
      .appliesWhen(appliesWhen);

    expect(policy.toRaw()).toEqual({
      name: 'http-errors',
      priority: 10,
      predicate: {
        nodes: [
          { tag: 'pred-and', val: [1, 2] },
          {
            tag: 'prop-eq',
            val: {
              propertyName: 'verb',
              value: { tag: 'text', val: 'get' },
            },
          },
          {
            tag: 'prop-gte',
            val: {
              propertyName: 'status-code',
              value: { tag: 'integer', val: 500n },
            },
          },
        ],
      },
      policy: {
        nodes: [
          {
            tag: 'filtered-on',
            val: {
              predicate: {
                nodes: [
                  {
                    tag: 'prop-matches',
                    val: {
                      propertyName: 'error-type',
                      pattern: 'timeout*',
                    },
                  },
                ],
              },
              inner: 1,
            },
          },
          {
            tag: 'jitter',
            val: {
              factor: 0.25,
              inner: 2,
            },
          },
          {
            tag: 'count-box',
            val: {
              maxRetries: 3,
              inner: 3,
            },
          },
          {
            tag: 'exponential',
            val: {
              baseDelay: 1_000_000_000n,
              factor: 2,
            },
          },
        ],
      },
    });
  });

  it('converts duration and property helpers ergonomically', () => {
    expect(Duration.minutes(2).toRaw()).toBe(120_000_000_000n);
    expect(toRawDuration(42)).toBe(42n);
    expect(Props.entries({ [Props.verb]: 'invoke', [Props.statusCode]: 503 })).toEqual([
      ['verb', { tag: 'text', val: 'invoke' }],
      ['status-code', { tag: 'integer', val: 503n }],
    ]);
    expect(toRawPredicateValue(true)).toEqual({ tag: 'boolean', val: true });
  });

  it('validates unsafe integer and policy inputs', () => {
    expect(() => toRawPredicateValue(1.5)).toThrow('safe integer');
    expect(() => toRawPredicateValue(1n << 63n)).toThrow('signed 64-bit');
    expect(() => toRawDuration(-1)).toThrow('non-negative');
    expect(() => toRawDuration(Number.MAX_SAFE_INTEGER + 1)).toThrow('safe integer');
    expect(() => Policy.exponential(Duration.seconds(1), 0).toRaw()).toThrow('greater than 0');
    expect(() => Policy.immediate().withJitter(-0.1).toRaw()).toThrow('greater than or equal to 0');
    expect(() =>
      Policy.immediate().clamp(Duration.seconds(2), Duration.seconds(1)).toRaw(),
    ).toThrow('less than or equal');
    expect(() =>
      NamedPolicy.named('bad-priority', Policy.immediate()).priority(-1).toRaw(),
    ).toThrow('32-bit integer');
  });

  it('passes raw named policies through untouched', () => {
    const raw = {
      name: 'raw',
      priority: 1,
      predicate: { nodes: [{ tag: 'pred-true' as const }] },
      policy: { nodes: [{ tag: 'immediate' as const }] },
    };

    expect(toRawNamedPolicy(raw)).toBe(raw);
  });

  it('rejects spoofed branded objects that do not implement toRaw', () => {
    const spoofedNamedPolicy = {
      [Symbol.for('golem.retry.named-policy')]: true,
    } as unknown as Parameters<typeof toRawNamedPolicy>[0];

    expect(() =>
      toRawDuration({ [Symbol.for('golem.retry.duration')]: true } as unknown as bigint),
    ).toThrow('safe integer');

    expect(toRawNamedPolicy(spoofedNamedPolicy)).toBe(spoofedNamedPolicy);
  });
});

describe('retry host helpers', () => {
  let getRetryPolicyByNameMock: ReturnType<typeof vi.fn>;
  let setRetryPolicyMock: ReturnType<typeof vi.fn>;
  let removeRetryPolicyMock: ReturnType<typeof vi.fn>;

  beforeEach(() => {
    getRetryPolicyByNameMock = vi.fn();
    setRetryPolicyMock = vi.fn();
    removeRetryPolicyMock = vi.fn();
  });

  afterEach(() => {
    vi.doUnmock('golem:api/retry@1.5.0');
    vi.resetModules();
  });

  async function loadRetryModule() {
    vi.resetModules();
    vi.doMock('golem:api/retry@1.5.0', () => ({
      getRetryPolicies: vi.fn(() => []),
      getRetryPolicyByName: getRetryPolicyByNameMock,
      resolveRetryPolicy: vi.fn(),
      setRetryPolicy: setRetryPolicyMock,
      removeRetryPolicy: removeRetryPolicyMock,
    }));

    return import('../src/host/retry');
  }

  it('setRetryPolicy accepts a NamedPolicy builder', async () => {
    const retry = await loadRetryModule();
    const policy = NamedPolicy.named('builder', Policy.immediate());

    retry.setRetryPolicy(policy);

    expect(setRetryPolicyMock).toHaveBeenCalledWith({
      name: 'builder',
      priority: 0,
      predicate: { nodes: [{ tag: 'pred-true' }] },
      policy: { nodes: [{ tag: 'immediate' }] },
    });
  });

  it('useRetryPolicy restores the previous raw policy on drop', async () => {
    const previous = {
      name: 'builder',
      priority: 1,
      predicate: { nodes: [{ tag: 'pred-true' as const }] },
      policy: { nodes: [{ tag: 'never' as const }] },
    };

    getRetryPolicyByNameMock.mockReturnValue(previous);

    const retry = await loadRetryModule();
    const guard = retry.useRetryPolicy(NamedPolicy.named('builder', Policy.immediate()));

    expect(setRetryPolicyMock).toHaveBeenNthCalledWith(1, {
      name: 'builder',
      priority: 0,
      predicate: { nodes: [{ tag: 'pred-true' }] },
      policy: { nodes: [{ tag: 'immediate' }] },
    });

    guard.drop();

    expect(setRetryPolicyMock).toHaveBeenNthCalledWith(2, previous);
    expect(removeRetryPolicyMock).not.toHaveBeenCalled();
  });

  it('withRetryPolicy removes newly added policies after the callback completes', async () => {
    getRetryPolicyByNameMock.mockReturnValue(undefined);

    const retry = await loadRetryModule();
    const result = retry.withRetryPolicy(NamedPolicy.named('temporary', Policy.immediate()), () => {
      return 'ok';
    });

    expect(result).toBe('ok');
    expect(removeRetryPolicyMock).toHaveBeenCalledWith('temporary');
  });
});

File: sdks/go/golem/retry/retry_test.go
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package retry

import (
	"math"
	"reflect"
	"strings"
	"testing"
	"time"

	apiRetry "github.com/golemcloud/golem/sdks/go/golem/internal/wit/golem_api_retry"
	witTypes "go.bytecodealliance.org/pkg/wit/types"
)

// TestPolicyLoweringStructure — a nested policy lowers to the expected flattened
// node array with root at index 0 and correct child indices (mirrors the Rust
// reference test): CountBox(0) -> FilteredOn(1) -> ClampDelay(2) -> Exponential(3).
func TestPolicyLoweringStructure(t *testing.T) {
	pol := Exponential(200*time.Millisecond, 2.0).
		Clamp(100*time.Millisecond, 5*time.Second).
		OnlyWhen(StatusCode.OneOf(502, 503, 504)).
		MaxRetries(3)

	raw, err := pol.lower()
	if err != nil {
		t.Fatalf("lower: %v", err)
	}
	if len(raw.Nodes) != 4 {
		t.Fatalf("nodes len = %d, want 4", len(raw.Nodes))
	}
	if raw.Nodes[0].Tag() != apiRetry.PolicyNodeCountBox {
		t.Fatalf("nodes[0] tag = %d, want CountBox", raw.Nodes[0].Tag())
	}
	if cb := raw.Nodes[0].CountBox(); cb.MaxRetries != 3 || cb.Inner != 1 {
		t.Fatalf("nodes[0] = %+v, want {MaxRetries:3 Inner:1}", cb)
	}
	if raw.Nodes[1].Tag() != apiRetry.PolicyNodeFilteredOn || raw.Nodes[1].FilteredOn().Inner != 2 {
		t.Fatalf("nodes[1] = %+v, want FilteredOn{Inner:2}", raw.Nodes[1])
	}
	if raw.Nodes[2].Tag() != apiRetry.PolicyNodeClampDelay {
		t.Fatalf("nodes[2] tag = %d, want ClampDelay", raw.Nodes[2].Tag())
	}
	if cl := raw.Nodes[2].ClampDelay(); cl.Inner != 3 || cl.MinDelay != uint64(100*time.Millisecond) || cl.MaxDelay != uint64(5*time.Second) {
		t.Fatalf("nodes[2] = %+v", cl)
	}
	if raw.Nodes[3].Tag() != apiRetry.PolicyNodeExponential {
		t.Fatalf("nodes[3] tag = %d, want Exponential", raw.Nodes[3].Tag())
	}
	if e := raw.Nodes[3].Exponential(); e.BaseDelay != uint64(200*time.Millisecond) || e.Factor != 2.0 {
		t.Fatalf("nodes[3] = %+v", e)
	}

	// The embedded predicate (from OnlyWhen) is a self-contained PropIn.
	pred := raw.Nodes[1].FilteredOn().Predicate
	if len(pred.Nodes) != 1 || pred.Nodes[0].Tag() != apiRetry.PredicateNodePropIn {
		t.Fatalf("embedded predicate = %+v, want single PropIn", pred.Nodes)
	}
	if in := pred.Nodes[0].PropIn(); in.PropertyName != "status-code" || len(in.Values) != 3 {
		t.Fatalf("PropIn = %+v", in)
	}
}

// TestPredicateLoweringStructure — And lowers to a combinator at root referencing
// its two operands by index.
func TestPredicateLoweringStructure(t *testing.T) {
	pred := Verb.Eq("GET").And(StatusCode.Gte(500))
	raw, err := pred.lower()
	if err != nil {
		t.Fatalf("lower: %v", err)
	}
	if len(raw.Nodes) != 3 {
		t.Fatalf("nodes len = %d, want 3", len(raw.Nodes))
	}
	if raw.Nodes[0].Tag() != apiRetry.PredicateNodePredAnd {
		t.Fatalf("nodes[0] tag = %d, want PredAnd", raw.Nodes[0].Tag())
	}
	if a := raw.Nodes[0].PredAnd(); a.F0 != 1 || a.F1 != 2 {
		t.Fatalf("PredAnd = %+v, want {1,2}", a)
	}
	if raw.Nodes[1].Tag() != apiRetry.PredicateNodePropEq || raw.Nodes[1].PropEq().PropertyName != "verb" {
		t.Fatalf("nodes[1] = %+v", raw.Nodes[1])
	}
	if raw.Nodes[2].Tag() != apiRetry.PredicateNodePropGte {
		t.Fatalf("nodes[2] tag = %d, want PropGte", raw.Nodes[2].Tag())
	}
	if v := raw.Nodes[1].PropEq().Value; v.Tag() != apiRetry.PredicateValueText || v.Text() != "GET" {
		t.Fatalf("verb value = %+v", v)
	}
	if v := raw.Nodes[2].PropGte().Value; v.Tag() != apiRetry.PredicateValueInteger || v.Integer() != 500 {
		t.Fatalf("status value = %+v", v)
	}
}

// TestNamedDefaults — Named defaults to priority 0 and an always-applies predicate.
func TestNamedDefaults(t *testing.T) {
	raw, err := Named("p", Immediate()).lower()
	if err != nil {
		t.Fatalf("lower: %v", err)
	}
	if raw.Name != "p" || raw.Priority != 0 {
		t.Fatalf("name/priority = %q/%d", raw.Name, raw.Priority)
	}
	if len(raw.Predicate.Nodes) != 1 || raw.Predicate.Nodes[0].Tag() != apiRetry.PredicateNodePredTrue {
		t.Fatalf("default predicate = %+v, want single PredTrue", raw.Predicate.Nodes)
	}
}

func TestValidationErrors(t *testing.T) {
	cases := []struct {
		name string
		np   NamedPolicy
		want string
	}{
		{"zero policy", Named("x", Policy{}), "zero) Policy"},
		{"exponential factor 0", Named("x", Exponential(time.Second, 0)), "factor must be finite and > 0"},
		{"exponential factor inf", Named("x", Exponential(time.Second, math.Inf(1))), "factor must be finite"},
		{"jitter negative", Named("x", Immediate().WithJitter(-1)), "factor must be finite and >= 0"},
		{"clamp min>max", Named("x", Immediate().Clamp(2*time.Second, time.Second)), "must be <= max"},
		{"negative duration", Named("x", Periodic(-time.Second)), "must be non-negative"},
		{"bad predicate value", Named("x", Immediate().OnlyWhen(StatusCode.Eq(1.5))), "unsupported predicate value type"},
		{"integer overflow", Named("x", Immediate().OnlyWhen(StatusCode.Eq(uint64(math.MaxInt64) + 1))), "overflows int64"},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			err := c.np.Validate()
			if err == nil {
				t.Fatalf("expected error containing %q, got nil", c.want)
			}
			if !strings.Contains(err.Error(), c.want) {
				t.Fatalf("error %q does not contain %q", err.Error(), c.want)
			}
		})
	}
}

// TestValidPolicyValidates — a well-formed policy passes Validate.
func TestValidPolicyValidates(t *testing.T) {
	np := Named("ok", Exponential(time.Second, 2).WithJitter(0.5).MaxRetries(5)).
		WithPriority(7).
		AppliesWhen(Verb.Eq("GET").Or(Verb.Eq("HEAD")))
	if err := np.Validate(); err != nil {
		t.Fatalf("Validate: %v", err)
	}
}

// TestRoundTrip — lower -> decode -> lower yields an identical node graph, for
// both the policy and its embedded/applicability predicates.
func TestRoundTrip(t *testing.T) {
	np := Named("rt", Exponential(200*time.Millisecond, 2).
		Clamp(0, time.Second).
		AddDelay(50*time.Millisecond).
		OnlyWhen(ErrorType.Contains("timeout")).
		MaxRetries(3).
		AndThen(Periodic(time.Second).Within(time.Minute))).
		WithPriority(9).
		AppliesWhen(StatusCode.Gte(500).And(URIHost.MatchesGlob("*.example.com")).Not())

	raw1, err := np.lower()
	if err != nil {
		t.Fatalf("lower1: %v", err)
	}
	decoded, err := decodeNamed(raw1)
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	raw2, err := decoded.lower()
	if err != nil {
		t.Fatalf("lower2: %v", err)
	}
	if !reflect.DeepEqual(raw1.Policy.Nodes, raw2.Policy.Nodes) {
		t.Fatalf("policy round-trip mismatch:\n %+v\n %+v", raw1.Policy.Nodes, raw2.Policy.Nodes)
	}
	if !reflect.DeepEqual(raw1.Predicate.Nodes, raw2.Predicate.Nodes) {
		t.Fatalf("predicate round-trip mismatch:\n %+v\n %+v", raw1.Predicate.Nodes, raw2.Predicate.Nodes)
	}
	if raw2.Name != "rt" || raw2.Priority != 9 {
		t.Fatalf("named fields lost: %q/%d", raw2.Name, raw2.Priority)
	}
}

func TestDecodeGuards(t *testing.T) {
	// empty node list
	if _, err := decodePolicy(apiRetry.RetryPolicy{}); err == nil || !strings.Contains(err.Error(), "empty policy node list") {
		t.Fatalf("empty: %v", err)
	}
	// dangling index: single CountBox pointing at index 5
	dangling := apiRetry.RetryPolicy{Nodes: []apiRetry.PolicyNode{
		apiRetry.MakePolicyNodeCountBox(apiRetry.CountBoxConfig{MaxRetries: 1, Inner: 5}),
	}}
	if _, err := decodePolicy(dangling); err == nil || !strings.Contains(err.Error(), "out of range") {
		t.Fatalf("dangling: %v", err)
	}
	// cycle: node0 -> node1 -> node0
	cyclic := apiRetry.RetryPolicy{Nodes: []apiRetry.PolicyNode{
		apiRetry.MakePolicyNodeCountBox(apiRetry.CountBoxConfig{MaxRetries: 1, Inner: 1}),
		apiRetry.MakePolicyNodeCountBox(apiRetry.CountBoxConfig{MaxRetries: 1, Inner: 0}),
	}}
	if _, err := decodePolicy(cyclic); err == nil || !strings.Contains(err.Error(), "cycle") {
		t.Fatalf("cycle: %v", err)
	}
	// predicate combinator with a dangling operand
	badPred := apiRetry.RetryPredicate{Nodes: []apiRetry.PredicateNode{
		apiRetry.MakePredicateNodePredAnd(witTypes.Tuple2[int32, int32]{F0: 1, F1: 9}),
		apiRetry.MakePredicateNodePredTrue(),
	}}
	if _, err := decodePredicate(badPred); err == nil || !strings.Contains(err.Error(), "out of range") {
		t.Fatalf("bad predicate: %v", err)
	}
}
